crates/cargo-codspeed/src/build.rs

```diff
@@ -1,168 +1,187 @@
+        if self.workspace {
+            cargo.arg("--workspace");
+        }
+
+        if !self.package.is_empty() {
+            self.package.iter().for_each(|p| {
+                cargo.arg("--package").arg(p);
+            });
+        }
+
+        if !self.exclude.is_empty() {
+            self.exclude.iter().for_each(|p| {
+                cargo.arg("--exclude").arg(p);
+            });
+        }
+    }
 }
 
 pub fn build_benches(
-    ws: &Workspace,
-    selected_benches: Option<Vec<String>>,
-    packages: Packages,
+    metadata: &Metadata,
+    filters: Filters,
     features: Option<Vec<String>>,
     profile: String,
+    quiet: bool,
 ) -> Result<()> {
-    let packages_to_build = packages.get_packages(ws)?;
-    let mut benches_to_build = vec![];
-    for package in packages_to_build.iter() {
-        let benches = package
-            .targets()
-            .iter()
-            .filter(|t| t.is_bench())
-            .collect_vec();
-        benches_to_build.extend(
-            benches
-                .into_iter()
-                .map(|t| BenchToBuild { package, target: t }),
+    let built_benches = BuildOptions {
+        filters,
+        features: &features,
+        profile: &profile,
+    }
+    .build(metadata, quiet)?;
+
+    if built_benches.is_empty() {
+        bail!(
+            "No benchmark target found. \
+            Please add a benchmark target to your Cargo.toml"
         );
     }
 
-    let all_benches_count = benches_to_build.len();
-
-    let benches_to_build = if let Some(selected_benches) = selected_benches {
-        benches_to_build
-            .into_iter()
-            .filter(|t| selected_benches.contains(&t.target.name().to_string()))
-            .collect_vec()
-    } else {
-        benches_to_build
-    };
-
-    let actual_benches_count = benches_to_build.len();
-
-    ws.gctx().shell().set_verbosity(Verbosity::Normal);
-    ws.gctx().shell().status_with_color(
-        "Collected",
-        format!(
-            "{} benchmark suite(s) to build{}",
-            benches_to_build.len(),
-            if all_benches_count > actual_benches_count {
-                format!(
-                    " ({} filtered out)",
-                    all_benches_count - actual_benches_count
-                )
-            } else {
-                "".to_string()
-            }
-        ),
-        &style::TITLE,
-    )?;
-
-    let context = ws.gctx();
-    let codspeed_root_target_dir = get_codspeed_target_dir(ws);
-    // Create and clear packages target directories
-    for package in packages_to_build.iter() {
-        let package_target_dir = codspeed_root_target_dir.join(package.name());
-        create_dir_all(&package_target_dir)?;
+    let codspeed_target_dir = get_codspeed_target_dir(metadata);
+    let built_bench_count = built_benches.len();
+
+    // Create and clear packages codspeed target directories
+    for built_bench in built_benches.iter() {
```

Comment:
We can probably find a way to avoid clearing the same directory multiple times here. Either a kind of set or `.map` and `.unique`?

Reply:
(it will also make the purpose of this first loop clearer)